Workflow failure: FIPS IT - KcSamlEncryptedIdTest#testEncryptedElementIsReadableInDeprecatedMode

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

ci

### Describe the bug

[FIPS IT](https://github.com/keycloak/keycloak/actions/runs/7563414138/job/20596089969)

```
Error: OR]   Run 3: KcSamlEncryptedIdTest>AbstractKcSamlEncryptedElementsTest.testEncryptedElementIsReadableInDeprecatedMode:98->AbstractKcSamlEncryptedElementsTest.sendDocumentWithEncryptedElement:158 
-- Expected: response status code matches is <200>
--      but: was <HttpResponseProxy{HTTP/1.1 400 Bad Request [Set-Cookie: KEYCLOAK_LOCALE=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/; HttpOnly, Set-Cookie: KC_RESTART=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/; HttpOnly, Set-Cookie: KC_AUTH_STATE=; Version=1; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/, Set-Cookie: KEYCLOAK_IDENTITY=***; Version=1; Path=/auth/realms/consumer/; SameSite=None; Secure; HttpOnly, Set-Cookie: KEYCLOAK_IDENTITY_LEGACY=***; Version=1; Path=/auth/realms/consumer/; HttpOnly, Set-Cookie: KEYCLOAK_SESSION=consumer/59efaf1d-bb4e-4f89-b40b-d7615316da33/e2698a21-a1e1-4ca8-a6c9-db15e38dd2b1; Version=1; Expires=Thu, 18-Jan-2024 10:33:04 GMT; Max-Age=36000; Path=/auth/realms/consumer/; SameSite=None; Secure, Set-Cookie: KEYCLOAK_SESSION_LEGACY=consumer/59efaf1d-bb4e-4f89-b40b-d7615316da33/e2698a21-a1e1-4ca8-a6c9-db15e38dd2b1; Version=1; Expires=Thu, 18-Jan-2024 10:33:04 GMT; Max-Age=36000; Path=/auth/realms/consumer/, Set-Cookie: KEYCLOAK_REMEMBER_ME=; Version=1; Comment=Expiring cookie; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Max-Age=0; Path=/auth/realms/consumer/; HttpOnly, content-length: 1743, Cache-Control: no-cache, Content-Language: en, Content-Security-Policy: frame-src 'self'; frame-ancestors 'self'; object-src 'none';, Content-Type: text/html;charset=utf-8, Referrer-Policy: no-referrer, Strict-Transport-Security: max-age=31536000; includeSubDomains, X-Content-Type-Options: nosniff, X-Frame-Options: SAMEORIGIN, X-Robots-Tag: none, X-XSS-Protection: 1; mode=block] ResponseEntityProxy{[Content-Type: text/html;charset=utf-8,Content-Length: 1743,Chunked: false]}}> with entity <!DOCTYPE html>
-- <html class="login-pf">
-- 
-- <head>
--     <meta charset="utf-8">
--     <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
--     <meta name="robots" content="noindex, nofollow">
-- 
--             <meta name="viewport" content="width=device-width,initial-scale=1"/>
--     <title>Sign in to consumer</title>
--     <link rel="icon" href="/auth/resources/5uoby/login/keycloak/img/favicon.ico" />
--             <link href="/auth/resources/5uoby/common/keycloak/node_modules/@patternfly/patternfly/patternfly.min.css" rel="stylesheet" />
--             <link href="/auth/resources/5uoby/common/keycloak/node_modules/patternfly/dist/css/patternfly.min.css" rel="stylesheet" />
--             <link href="/auth/resources/5uoby/common/keycloak/node_modules/patternfly/dist/css/patternfly-additions.min.css" rel="stylesheet" />
--             <link href="/auth/resources/5uoby/common/keycloak/lib/pficon/pficon.css" rel="stylesheet" />
--             <link href="/auth/resources/5uoby/login/keycloak/css/login.css" rel="stylesheet" />
-- </head>
-- 
-- <body class="">
-- <div class="login-pf-page">
--     <div id="kc-header" class="login-pf-page-header">
--         <div id="kc-header-wrapper"
--              class="">consumer</div>
--     </div>
--     <div class="card-pf">
--         <header class="login-pf-header">
--                 <h1 id="kc-page-title">        We are sorry...
-- </h1>
--       </header>
--       <div id="kc-content">
--         <div id="kc-content-wrapper">
-- 
-- 
--         <div id="kc-error-message">
--             <p class="instruction">Failed to process response</p>
--                     <p><a id="backToApplication" href="https://localhost:8543/sales-post">? Back to Application</a></p>
--         </div>
-- 
-- 
-- 
--         </div>
--       </div>
-- 
--     </div>
--   </div>
-- </body>
-- </html>
-- 
-- [INFO] 
-- [INFO] 
Error: OR] Tests run: 467, Failures: 1, Errors: 0, Skipped: 7
```

### Version

latest

### Expected behavior

-

### Actual behavior

-

### How to Reproduce?

-

### Anything else?

_No response_

It seems the cause of these failures is shown here:

```
KcSamlEncryptedIdTest ++ 00:33:04,457 INFO  [org.keycloak.testsuite.util.SamlClient] Executing HTTP request to https://localhost:8543/auth/realms/consumer/broker/after-first-broker-login?session_code=ZT9jLj4xMB96uIIQ60W35Yk67rsxQH31PYB7eLyewiQ&client_id=http%3A%2F%2Flocalhost%3A8280%2Fsales-post%2F.dot%2Fted&tab_id=E5yB1mAac2c
KcSamlEncryptedIdTest ++ 2024-01-18 00:33:04,465 ERROR [org.keycloak.protocol.saml.SamlProtocol] (executor-thread-2) failed: org.keycloak.saml.common.exceptions.ProcessingException: javax.xml.crypto.dsig.XMLSignatureException: PL00100: Signing Process Failure:
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.processing.api.saml.v2.sig.SAML2Signature.signSAMLDocument(SAML2Signature.java:158)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.BaseSAML2BindingBuilder.signDocument(BaseSAML2BindingBuilder.java:321)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.BaseSAML2BindingBuilder$BasePostBindingBuilder.<init>(BaseSAML2BindingBuilder.java:174)
KcSamlEncryptedIdTest ++ 	at org.keycloak.protocol.saml.JaxrsSAML2BindingBuilder$PostBindingBuilder.<init>(JaxrsSAML2BindingBuilder.java:50)
KcSamlEncryptedIdTest ++ 	at org.keycloak.protocol.saml.JaxrsSAML2BindingBuilder.postBinding(JaxrsSAML2BindingBuilder.java:122)
KcSamlEncryptedIdTest ++ 	at org.keycloak.protocol.saml.SamlProtocol.buildAuthenticatedResponse(SamlProtocol.java:564)
KcSamlEncryptedIdTest ++ 	at org.keycloak.protocol.saml.SamlProtocol.authenticated(SamlProtocol.java:555)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.managers.AuthenticationManager.redirectAfterSuccessfulFlow(AuthenticationManager.java:989)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.managers.AuthenticationManager.redirectAfterSuccessfulFlow(AuthenticationManager.java:937)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.managers.AuthenticationManager.finishedRequiredActions(AuthenticationManager.java:1090)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService.finishBrokerAuthentication(IdentityBrokerService.java:851)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService.afterPostBrokerLoginFlowSuccess(IdentityBrokerService.java:805)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService.finishOrRedirectToPostBrokerLogin(IdentityBrokerService.java:743)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService.afterFirstBrokerLogin(IdentityBrokerService.java:730)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService.afterFirstBrokerLogin(IdentityBrokerService.java:645)
KcSamlEncryptedIdTest ++ 	at org.keycloak.services.resources.IdentityBrokerService$quarkusrestinvoker$afterFirstBrokerLogin_2875cf30d412fd2f366643b0cfdc1bbfd26a9393.invoke(Unknown Source)
KcSamlEncryptedIdTest ++ 	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
KcSamlEncryptedIdTest ++ 	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
KcSamlEncryptedIdTest ++ 	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
KcSamlEncryptedIdTest ++ 	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
KcSamlEncryptedIdTest ++ 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
KcSamlEncryptedIdTest ++ 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
KcSamlEncryptedIdTest ++ 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
KcSamlEncryptedIdTest ++ 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
KcSamlEncryptedIdTest ++ 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
KcSamlEncryptedIdTest ++ 	at java.base/java.lang.Thread.run(Thread.java:840)
KcSamlEncryptedIdTest ++ Caused by: javax.xml.crypto.dsig.XMLSignatureException: PL00100: Signing Process Failure:
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.common.DefaultPicketLinkLogger.signatureError(DefaultPicketLinkLogger.java:185)
KcSamlEncryptedIdTest ++ 	... 26 more
KcSamlEncryptedIdTest ++ Caused by: javax.xml.crypto.dsig.XMLSignatureException: java.security.InvalidKeyException: Cannot initialize for signing: Attempt to sign/verify with RSA modulus already used for encrypt/decrypt.
KcSamlEncryptedIdTest ++ 	at java.xml.crypto/org.jcp.xml.dsig.internal.dom.DOMXMLSignature.sign(DOMXMLSignature.java:405)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.processing.core.util.XMLSignatureUtil.signImpl(XMLSignatureUtil.java:713)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.processing.core.util.XMLSignatureUtil.sign(XMLSignatureUtil.java:393)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.processing.api.saml.v2.sig.SAML2Signature.sign(SAML2Signature.java:139)
KcSamlEncryptedIdTest ++ 	at org.keycloak.saml.processing.api.saml.v2.sig.SAML2Signature.signSAMLDocument(SAML2Signature.java:156)
KcSamlEncryptedIdTest ++ 	... 25 more
KcSamlEncryptedIdTest ++ Caused by: java.security.InvalidKeyException: Cannot initialize for signing: Attempt to sign/verify with RSA modulus already used for encrypt/decrypt.
KcSamlEncryptedIdTest ++ 	at org.bouncycastle.jcajce.provider.BaseSignature.engineInitSign(Unknown Source)
KcSamlEncryptedIdTest ++ 	at java.base/java.security.Signature$Delegate.tryOperation(Signature.java:1327)
KcSamlEncryptedIdTest ++ 	at java.base/java.security.Signature$Delegate.chooseProvider(Signature.java:1276)
KcSamlEncryptedIdTest ++ 	at java.base/java.security.Signature$Delegate.engineInitSign(Signature.java:1373)
KcSamlEncryptedIdTest ++ 	at java.base/java.security.Signature.initSign(Signature.java:635)
KcSamlEncryptedIdTest ++ 	at java.xml.crypto/org.jcp.xml.dsig.internal.dom.DOMSignatureMethod.sign(DOMSignatureMethod.java:335)
KcSamlEncryptedIdTest ++ 	at java.xml.crypto/org.jcp.xml.dsig.internal.dom.DOMXMLSignature.sign(DOMXMLSignature.java:402)
KcSamlEncryptedIdTest ++ 	... 29 more
KcSamlEncryptedIdTest ++ Caused by: org.bouncycastle.crypto.IllegalKeyException: Attempt to sign/verify with RSA modulus already used for encrypt/decrypt.
KcSamlEncryptedIdTest ++ 	at org.bouncycastle.crypto.fips.FipsRSA$SignatureOperatorFactory.createSigner(Unknown Source)
KcSamlEncryptedIdTest ++ 	at org.bouncycastle.crypto.fips.FipsRSA$SignatureOperatorFactory.createSigner(Unknown Source)
KcSamlEncryptedIdTest ++ 	at org.bouncycastle.jcajce.provider.ProvRSA$AdaptiveSignatureOperatorFactory.createSigner(Unknown Source)
KcSamlEncryptedIdTest ++ 	at org.bouncycastle.jcajce.provider.ProvRSA$AdaptiveSignatureOperatorFactory.createSigner(Unknown Source)
KcSamlEncryptedIdTest ++ 	... 36 more
KcSamlEncryptedIdTest ++ 
```

@mposolda @rmartinc Does it ring any bell?

--- a/docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc
+++ b/docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc
@@ -220,3 +220,7 @@ PUT /admin/realms/{realm}/users/{id}/execute-actions-email
 
 ["VERIFY_EMAIL"]
 ----
+
+= Removal of the deprecated mode for SAML encryption
+
+The compatibility mode for SAML encryption introduced in version 21 is now removed. The system property `keycloak.saml.deprecated.encryption` is not managed anymore by the server. The clients which still used the old signing key for encryption should update it from the new IDP configuration metadata.

--- a/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java
+++ b/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java
@@ -90,9 +90,7 @@ import jakarta.ws.rs.core.UriInfo;
 import javax.xml.namespace.QName;
 import java.io.IOException;
 import java.security.Key;
-import java.security.PrivateKey;
 import java.security.cert.X509Certificate;
-import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.LinkedList;
@@ -156,9 +154,6 @@ public class SAMLEndpoint {
     private final ClientConnection clientConnection;
 
     private final HttpHeaders headers;
-
-    public static final String ENCRYPTION_DEPRECATED_MODE_PROPERTY = "keycloak.saml.deprecated.encryption";
-    private final boolean DEPRECATED_ENCRYPTION = Boolean.getBoolean(ENCRYPTION_DEPRECATED_MODE_PROPERTY);
 
 
     public SAMLEndpoint(KeycloakSession session, SAMLIdentityProvider provider, SAMLIdentityProviderConfig config, IdentityProvider.AuthenticationCallback callback, DestinationValidator destinationValidator) {
@@ -460,17 +455,6 @@ public class SAMLEndpoint {
                 if (assertionIsEncrypted) {
                     try {
                         XMLEncryptionUtil.DecryptionKeyLocator decryptionKeyLocator = new SAMLDecryptionKeysLocator(session, realm, config.getEncryptionAlgorithm());
-                        /* This code is deprecated and will be removed in Keycloak 24 */
-                        if (DEPRECATED_ENCRYPTION) {
-                            KeyManager.ActiveRsaKey keys = session.keys().getActiveRsaKey(realm);
-                            final XMLEncryptionUtil.DecryptionKeyLocator tmp = decryptionKeyLocator;
-                            decryptionKeyLocator = data -> {
-                                List<PrivateKey> result = new ArrayList<>(tmp.getKeys(data));
-                                result.add(keys.getPrivateKey());
-                                return result;
-                            };
-                        }
-                        /* End of deprecated code */
                         assertionElement = AssertionUtil.decryptAssertion(responseType, decryptionKeyLocator);
                     } catch (ProcessingException ex) {
                         logger.warnf(ex, "Not possible to decrypt SAML assertion. Please check realm keys of usage ENC in the realm '%s' and make sure there is a key able to decrypt the assertion encrypted by identity provider '%s'", realm.getName(), config.getAlias());
@@ -518,17 +502,6 @@ public class SAMLEndpoint {
                 if (AssertionUtil.isIdEncrypted(responseType)) {
                     try {
                         XMLEncryptionUtil.DecryptionKeyLocator decryptionKeyLocator = new SAMLDecryptionKeysLocator(session, realm, config.getEncryptionAlgorithm());
-                        /* This code is deprecated and will be removed in Keycloak 24 */
-                        if (DEPRECATED_ENCRYPTION) {
-                            KeyManager.ActiveRsaKey keys = session.keys().getActiveRsaKey(realm);
-                            final XMLEncryptionUtil.DecryptionKeyLocator tmp = decryptionKeyLocator;
-                            decryptionKeyLocator = data -> {
-                                List<PrivateKey> result = new ArrayList<>(tmp.getKeys(data));
-                                result.add(keys.getPrivateKey());
-                                return result;
-                            };
-                        }
-                        /* End of deprecated code */
                         AssertionUtil.decryptId(responseType, decryptionKeyLocator);
                     } catch (ProcessingException ex) {
                         logger.warnf(ex, "Not possible to decrypt SAML encryptedId. Please check realm keys of usage ENC in the realm '%s' and make sure there is a key able to decrypt the encryptedId encrypted by identity provider '%s'", realm.getName(), config.getAlias());

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractKcSamlEncryptedElementsTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractKcSamlEncryptedElementsTest.java
@@ -48,7 +48,6 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.not;
-import static org.keycloak.broker.saml.SAMLEndpoint.ENCRYPTION_DEPRECATED_MODE_PROPERTY;
 import static org.keycloak.testsuite.broker.BrokerTestTools.getConsumerRoot;
 import static org.keycloak.testsuite.saml.AbstractSamlTest.SAML_CLIENT_ID_SALES_POST;
 import static org.keycloak.testsuite.util.Matchers.isSamlResponse;
@@ -84,24 +83,6 @@ public abstract class AbstractKcSamlEncryptedElementsTest extends AbstractBroker
             KeysMetadataRepresentation.KeyMetadataRepresentation activeSignatureKey = KeyUtils.findActiveSigningKey(adminClient.realm(bc.consumerRealmName()));
             assertThat(activeSignatureKey.getProviderId(), equalTo(sigProviderId));
             sendDocumentWithEncryptedElement(PemUtils.decodePublicKey(activeSignatureKey.getPublicKey()), XMLCipher.RSA_OAEP, null, null, false);
-        }
-
-        @Test
-        public void testEncryptedElementIsReadableInDeprecatedMode() throws ConfigurationException, ParsingException, ProcessingException {
-            try {
-                // Set flag that enabled deprecated mode for encryption
-                testingClient.server().run(session -> {
-                    System.setProperty(ENCRYPTION_DEPRECATED_MODE_PROPERTY, "true");
-                });
-                KeysMetadataRepresentation.KeyMetadataRepresentation activeSignatureKey = KeyUtils.findActiveSigningKey(adminClient.realm(bc.consumerRealmName()));
-                assertThat(activeSignatureKey.getProviderId(), equalTo(sigProviderId));
-                sendDocumentWithEncryptedElement(PemUtils.decodePublicKey(activeSignatureKey.getPublicKey()), XMLCipher.RSA_OAEP, null, null, true);
-            } finally {
-                // Clear flag
-                testingClient.server().run(session -> {
-                    System.clearProperty(ENCRYPTION_DEPRECATED_MODE_PROPERTY);
-                });
-            }
         }
 
         @Test

